File: ports/dotnet/src/RussianDocs.DocumentProcessing/Inference/Session.cs
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using RussianDocs.DocumentProcessing.Tensors;

namespace RussianDocs.DocumentProcessing.Inference;

/// <summary>Which device a session runs on. A string-backed enum, so it matches the wire values.</summary>
public enum Device
{
    Cpu,
    Gpu,
}

public static class DeviceNames
{
    public static string Wire(this Device device) => device == Device.Gpu ? "gpu" : "cpu";

    public static Device Parse(string? value) => value?.ToLowerInvariant() switch
    {
        "gpu" => Device.Gpu,
        "cpu" or null or "" => Device.Cpu,
        _ => throw new ArgumentException($"inference: unknown device \"{value}\""),
    };
}

/// <summary>
/// One ONNX Runtime session.
///
/// <para>
/// **The lock is per session and covers only <c>Run</c>.** It is not defensive tidiness: concurrent
/// <c>Run</c> calls on a single CUDA session were measured to destroy the process — Python saw
/// <c>cudaErrorIllegalAddress</c> after roughly 200 calls, and Go degraded from 6.6 s to over 600 s
/// for the same work, which for a service is indistinguishable from a hang. Holding it only around
/// <c>Run</c> is what preserves the speedup of the parallel groups, because those use five DIFFERENT
/// sessions and therefore five different locks.
/// </para>
///
/// <para>
/// CPU sessions deliberately do NOT lock. ONNX Runtime's CPU provider is re-entrant, and locking
/// there would serialise the quality group for no reason.
/// </para>
/// </summary>
public sealed class Session : IDisposable
{
    private readonly InferenceSession _session;
    private readonly object? _gate;
    private readonly Dictionary<string, NodeMetadata> _inputs;

    /// <summary>Output names in the order the model DECLARES them.</summary>
    public string[] OutputNames { get; }

    public Device Device { get; }

    public Session(string modelPath, Device device, int intraOpThreads)
    {
        Device = device;

        var options = new SessionOptions();
        try
        {
            if (device == Device.Gpu)
            {
                // Throws if the provider is unavailable, which the caller's [gpu, cpu] attempt loop
                // handles. What it CANNOT catch is a container started without --gpus, where the
                // provider segfaults instead — hence the device-node probe before we ever get here.
                options.AppendExecutionProvider_CUDA(0);
            }

            if (intraOpThreads > 0)
            {
                // Pinned for conformance runs. ONNX Runtime's CPU reductions split across threads,
                // so a different thread count legitimately shifts results by ~1e-6 — inside the
                // float tolerance, but enough to flip an argmax on near-equal values, which is an
                // exact-match failure with no float anywhere near it.
                options.IntraOpNumThreads = intraOpThreads;
            }

            _session = new InferenceSession(modelPath, options);
        }
        catch
        {
            options.Dispose();
            throw;
        }

        _inputs = _session.InputMetadata.ToDictionary(kv => kv.Key, kv => kv.Value);
        OutputNames = _session.OutputMetadata.Keys.ToArray();
        _gate = device == Device.Gpu ? new object() : null;
    }

    /// <summary>
    /// Runs the model. Inputs are positional; the caller supplies them in declaration order.
    ///
    /// <para>
    /// Each input is CAST to the dtype the model declares. The detectors want float32 and the OCR
    /// nets want uint8, and feeding the wrong one is not a graceful failure — ONNX Runtime either
    /// refuses or reinterprets the bytes.
    /// </para>
    /// </summary>
    public NdArray[] Run(IReadOnlyList<NdArray> inputs)
    {
        string[] names = _inputs.Keys.ToArray();
        if (inputs.Count != names.Length)
        {
            throw new ArgumentException(
                $"inference: model wants {names.Length} inputs, got {inputs.Count}");
        }

        var feeds = new List<NamedOnnxValue>(inputs.Count);
        var owned = new List<IDisposable>();
        try
        {
            for (int i = 0; i < inputs.Count; i++)
            {
                feeds.Add(ToNamedValue(names[i], inputs[i], _inputs[names[i]]));
            }

            IDisposableReadOnlyCollection<DisposableNamedOnnxValue> results;
            if (_gate is not null)
            {
                // Acquire, run, release — and read the results AFTER releasing, outside the lock.
                lock (_gate)
                {
                    results = _session.Run(feeds);
                }
            }
            else
            {
                results = _session.Run(feeds);
            }

            using (results)
            {
                // Collected BY DECLARED NAME, not by iteration order. DocTypeAngles has two outputs
                // whose meanings are not interchangeable — embeddings and an angle — and an
                // unexpected reordering must be visible rather than quietly swapping two heads.
                var byName = results.ToDictionary(v => v.Name, v => v);
                var outputs = new NdArray[OutputNames.Length];
                for (int i = 0; i < OutputNames.Length; i++)
                {
                    outputs[i] = FromOnnxValue(byName[OutputNames[i]]);
                }
                return outputs;
            }
        }
        finally
        {
            foreach (IDisposable d in owned)
            {
                d.Dispose();
            }
        }
    }

    private static NamedOnnxValue ToNamedValue(string name, NdArray array, NodeMetadata meta)
    {
        int[] shape = array.Shape;
        if (meta.ElementType == typeof(float))
        {
            float[] values = array.Dtype switch
            {
                Dtype.Float32 => array.AsFloat32().ToArray(),
                // uint8 to float32 without scaling: normalisation is baked into the graphs, and
                // dividing here would silently double-normalise.
                Dtype.UInt8 => ToFloats(array.AsUInt8()),
                _ => throw new InvalidOperationException(
                    $"inference: {name} wants float32, cannot cast from {array.Dtype}"),
            };
            return NamedOnnxValue.CreateFromTensor(name, new DenseTensor<float>(values, shape));
        }
        if (meta.ElementType == typeof(byte))
        {
            byte[] values = array.Dtype == Dtype.UInt8
                ? array.AsUInt8().ToArray()
                : throw new InvalidOperationException(
                    $"inference: {name} wants uint8, cannot cast from {array.Dtype}");
            return NamedOnnxValue.CreateFromTensor(name, new DenseTensor<byte>(values, shape));
        }
        throw new InvalidOperationException(
            $"inference: {name} has unsupported element type {meta.ElementType}");
    }

    private static float[] ToFloats(ReadOnlySpan<byte> bytes)
    {
        var values = new float[bytes.Length];
        for (int i = 0; i < bytes.Length; i++)
        {
            values[i] = bytes[i];
        }
        return values;
    }

    private static NdArray FromOnnxValue(DisposableNamedOnnxValue value)
    {
        if (value.ValueType != OnnxValueType.ONNX_TYPE_TENSOR)
        {
            throw new InvalidOperationException($"inference: {value.Name} is not a tensor");
        }

        if (value.ElementType == TensorElementType.Float)
        {
            var tensor = value.AsTensor<float>();
            return NdArray.FromFloat32(tensor.ToArray(), tensor.Dimensions.ToArray());
        }
        if (value.ElementType == TensorElementType.UInt8)
        {
            var tensor = value.AsTensor<byte>();
            return NdArray.FromUInt8(tensor.ToArray(), tensor.Dimensions.ToArray());
        }
        throw new InvalidOperationException(
            $"inference: {value.Name} has unsupported output type {value.ElementType}");
    }

    public void Dispose() => _session.Dispose();
}

File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/inference/Session.kt
package net.russiandocs.docproc.inference

import ai.onnxruntime.OnnxJavaType
import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import ai.onnxruntime.TensorInfo
import net.russiandocs.docproc.pipeline.Device
import net.russiandocs.docproc.tensors.Dtype
import net.russiandocs.docproc.tensors.NdArray
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * One ONNX Runtime session, plus the two rules that make it safe to share.
 *
 * ### The GPU lock
 *
 * Concurrent `run` calls on ONE CUDA session wedge the device. Measured, not theoretical: Python hit
 * `cudaErrorIllegalAddress` after roughly 200 calls, and the Go port degraded by more than 90x — 6.6 s
 * with the lock against over 600 s without, never finishing. The lock is held around `run` ONLY, so
 * five different models keep their parallelism, and the results are read after releasing it.
 *
 * **CPU sessions take no lock, deliberately.** ONNX Runtime's CPU path is re-entrant, and locking it
 * would serialise the quality group that exists precisely to run four classifiers at once.
 *
 * ### Casting to the declared dtype
 *
 * Each input is cast to what the model declares. The detectors want float32 and the OCR nets want
 * uint8; feeding the wrong one is not a graceful failure — ONNX Runtime either refuses or reinterprets
 * the bytes.
 */
public class Session(
    modelPath: String,
    public val device: Device,
    intraOpThreads: Int,
) : AutoCloseable {

    private val env: OrtEnvironment = OrtEnvironment.getEnvironment()
    private val session: OrtSession
    private val inputNames: List<String>
    private val inputTypes: Map<String, OnnxJavaType>

    /** Output names in the order the model DECLARES them. */
    public val outputNames: List<String>

    /** Null on CPU: see the type note. */
    private val gate: Any?

    init {
        val options = OrtSession.SessionOptions()
        try {
            if (device == Device.GPU) {
                // Throws if the provider is unavailable, which the caller's [gpu, cpu] attempt loop
                // handles. What it CANNOT catch is a container started without --gpus, where the
                // provider segfaults instead — hence the device-node probe before we ever get here.
                options.addCUDA(0)
            }
            if (intraOpThreads > 0) {
                // Pinned for conformance runs; see the note on the CLI's IntraOpThreads.
                options.setIntraOpNumThreads(intraOpThreads)
            }
            session = env.createSession(modelPath, options)
        } catch (e: Throwable) {
            options.close()
            throw e
        }

        // LinkedHashMap iteration order is the declaration order the binding reports, and that is what
        // "inputs are positional" below relies on.
        inputNames = session.inputNames.toList()
        inputTypes = session.inputInfo.mapValues { (_, info) ->
            (info.info as TensorInfo).type
        }
        outputNames = session.outputNames.toList()
        gate = if (device == Device.GPU) Any() else null
    }

    /**
     * Runs the model. Inputs are positional; the caller supplies them in declaration order.
     *
     * Outputs are collected **by declared name**, not by iteration order: `DocTypeAngles` has two
     * outputs whose meanings are not interchangeable — embeddings and an angle — and an unexpected
     * reordering must be visible rather than quietly swapping two heads.
     */
    public fun run(inputs: List<NdArray>): List<NdArray> {
        require(inputs.size == inputNames.size) {
            "inference: model wants ${inputNames.size} inputs, got ${inputs.size}"
        }

        val feeds = LinkedHashMap<String, OnnxTensor>(inputs.size)
        try {
            for ((i, name) in inputNames.withIndex()) {
                feeds[name] = toTensor(name, inputs[i], inputTypes.getValue(name))
            }

            val result = if (gate != null) {
                // Acquire, run, release — and read the results AFTER releasing, outside the lock.
                synchronized(gate) { session.run(feeds) }
            } else {
                session.run(feeds)
            }

            result.use { outputs ->
                return outputNames.map { name ->
                    fromTensor(outputs.get(name).orElseThrow {
                        IllegalStateException("inference: model declared output $name but did not " +
                            "return it")
                    } as OnnxTensor)
                }
            }
        } finally {
            feeds.values.forEach { it.close() }
        }
    }

    private fun toTensor(name: String, array: NdArray, want: OnnxJavaType): OnnxTensor {
        val shape = array.shape.map { it.toLong() }.toLongArray()
        return when (want) {
            OnnxJavaType.FLOAT -> {
                val values = when (array.dtype) {
                    Dtype.FLOAT32 -> array.asFloat32()
                    // uint8 to float32 WITHOUT scaling: normalisation is baked into the graphs, and
                    // dividing here would silently double-normalise.
                    Dtype.UINT8 -> FloatArray(array.length) { i ->
                        (array.data[i].toInt() and 0xff).toFloat()
                    }
                    Dtype.FLOAT64 -> {
                        val src = array.asFloat64()
                        FloatArray(src.size) { src[it].toFloat() }
                    }
                    else -> throw IllegalStateException(
                        "inference: $name wants float32, cannot cast from ${array.dtype}")
                }
                OnnxTensor.createTensor(env, java.nio.FloatBuffer.wrap(values), shape)
            }

            OnnxJavaType.UINT8 -> {
                val values = when (array.dtype) {
                    Dtype.UINT8 -> array.asUInt8()
                    else -> throw IllegalStateException(
                        "inference: $name wants uint8, cannot cast from ${array.dtype}")
                }
                // The UINT8 overload must be named explicitly: without it the binding infers INT8 from
                // a ByteBuffer and the model rejects the feed with a type error that names neither.
                OnnxTensor.createTensor(env, ByteBuffer.wrap(values), shape, OnnxJavaType.UINT8)
            }

            OnnxJavaType.INT64 -> {
                val values = when (array.dtype) {
                    Dtype.INT64 -> array.asInt64()
                    else -> throw IllegalStateException(
                        "inference: $name wants int64, cannot cast from ${array.dtype}")
                }
                OnnxTensor.createTensor(env, java.nio.LongBuffer.wrap(values), shape)
            }

            else -> throw IllegalStateException("inference: $name declares unsupported type $want")
        }
    }

    /**
     * Copies an output tensor into an [NdArray].
     *
     * Through the raw ByteBuffer rather than `getValue()`, and that is not an optimisation: `getValue()`
     * materialises a nested `Object[]` of boxed arrays — for a `[1, 8400, 26]` detector head that is
     * 8400 float arrays plus the boxing, per call, on the hot path. The buffer copy is one pass.
     */
    private fun fromTensor(tensor: OnnxTensor): NdArray {
        val info = tensor.info
        val shape = info.shape.map { it.toInt() }.toIntArray()
        val buffer = tensor.byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        return when (info.type) {
            OnnxJavaType.FLOAT -> NdArray(bytes, shape, Dtype.FLOAT32, 4)
            OnnxJavaType.DOUBLE -> NdArray(bytes, shape, Dtype.FLOAT64, 8)
            OnnxJavaType.UINT8, OnnxJavaType.INT8 -> NdArray(bytes, shape, Dtype.UINT8, 1)
            OnnxJavaType.INT64 -> NdArray(bytes, shape, Dtype.INT64, 8)
            else -> throw IllegalStateException(
                "inference: unsupported output type ${info.type}")
        }
    }

    override fun close() {
        session.close()
        // The environment is a process-wide singleton in the JVM binding — closing it here would tear
        // down every other session. Left alone deliberately.
    }
}
